### PR TITLE
Create sensitive and deleted media models for decisions

### LINKED ISSUE #4513: Create `MediaDecision` have no effect on deindexed actions
## Description

Currently, when trying to deindex a reported item, nothing happens beyond creating the record of the decision. In these cases, items are supposed to be deindexed and removed from the media table, saving the identifier in the corresponding deleted media table. This function is critical to remove inappropriate content from Openverse and decisions created while this is done will need to be backfilled.

## Reproduction

1. Report an image (or audio)
2. Select the action to deindex it via the Django admin
3. Verify the deleted images and see the table is empty
4. Also observe the image is still available in search and in its details endpoint

## Record

https://github.com/WordPress/openverse/assets/9145885/4899ccf5-1e29-40e6-b892-696af0abad74

## Additional context

Related to project #383.

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #4513 by @krysal 

## Description

This PR ensures that `SensitiveImage`/`SensitiveAudio` and `DeletedImage`/`DeletedAudio` models are created for every decision.

## Testing Instructions

0. Create a report for an image.
1. Take a deindexed-sensitive or deindexed-copyright action on the report.
   - You should see a new `DeletedImage` object.
   - The image should be deleted.
   - The report should be resolved with a linked decision.
2. Create a report for an image.
3. Take a marked-sensitive action on the report.
   - You should see a new `SenstiveMedia` object.
   - The image should be marked as sensitive.
   - The report should be resolved with a linked decision.

Repeat these steps for audio.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (`./ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`./ov just catalog/generate-docs media-props`
      for the catalog or `./ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by AetherUnbound — APPROVED** [review]
Makes sense, and looks good to me! I tested this locally and confirmed that the deindexed image gets deleted from the API database, and the sensitive image gets a record in the sensitive table.

File: api/api/management/commands/backfillmoderationdecision.py [hunk 1]
```diff
@@ -92,14 +92,18 @@ def handle(self, *args, **options):
...
                     MediaDecisionThrough(
                         media_obj_id=report.media_obj_id, decision=decision
                     )
                     for report, decision in zip(reports_chunk, decisions)
                 ]
             )
+            # Bulk create does not call ``save()`` so the mark-sensitive/deindex
+            # actions will not be undertaken.
+            for through in throughs:
+                through.perform_action()
```
Comment (sarayourfriend) [inline api/api/management/commands/backfillmoderationdecision.py:106]: This loop is a disappointing solution to the backfill, and somewhat defeats the purpose of the bulk creates above :confused: They're still useful as is, but this really goes against that effort by executing an individual query for each through instance regardless, especially because `perform_action` not only does the queries to create the deleted/sensitive model instances, but also relies on `self.decision.action`, which implicitly queries for `self.decision`, as well as `self.media_obj` which implicitly queries the entire media object. So the `perform_action` function is potentially very inefficient, especially when we consider it in light of the known case of bulk moderation actions, which must be able to operate on many records on an ongoing basis.

We can avoid the `self.media_obj` query within the method pretty easily. `self.decision.action` in this particular case (and potentially all cases where it would be relevant) is actually implicitly avoided (though honestly only accidentally, and is only an accident, not necessary) because the model will already have `decision` loaded from when it was created on line 98. I don't think we should rely on that implicit (accidental) fact to avoid the implicit query.

Instead, we should go ahead and make `perform_action` presume a bulk operation, and can be trivially adapted to do a single record more easily if we wanted (though I don't think it's necessary). Because we know decisions will be created in bulk, and because the bulk operation should perform the same as the non-bulk operation, I think we need to go ahead and solve that, potentially even down to the ES level. Unfortunately, that's a big task, and probably worth separating into a new PR instead of including all the fixes here.

Maybe something like this would work:

```
MediaObjectIdentifiers = namedtuple("identifier", "pk")
def decide(
    media_object_identifiers: list[MediaObjectIdentifiers],
    report_pks: list[int],
    action: str,
):
    # create decision
    # bulk update the reports
    # bulk create the through records (using `decision_id=decision.pk` rather than the full decision object)
    # bulk create the appropriate sensitive/deleted media instances based on the action
    # bulk update the filtered and non-filtered indexes using `Elasticsearch::bulk`
```

Anyway, I think either we should do that in this PR, or leave out this fix to the backfill and address it in a separate PR. I am strongly against introducing this kind of operation into the codebase, especially if it could accidentally be taken as an approved pattern to use when implementing the actual bulk decision creation.

File: api/test/unit/management/commands/test_backfillmoderationdecision.py [hunk 2]
```diff
@@ -78,12 +80,13 @@ def test_create_moderation_decision_for_reports(
...
     decision = MediaDecision.objects.first()
-    assert decision.media_objs.count() == 1
     assert decision.action == expected_action
     assert decision.moderator.username == username
 
+    assert MediaDecisionThrough.objects.filter(decision=decision).count() == 1
+
     decision_through = MediaDecisionThrough.objects.first()
-    assert decision_through.media_obj == report.media_obj
+    assert str(decision_through.media_obj_id) == report.media_obj_id
```
Comment (sarayourfriend) [inline api/test/unit/management/commands/test_backfillmoderationdecision.py:89]: Aren't these statements equivalent? Why the change here? When would they not be equivalent? Django models essentially compare equality through `self.pk == other.pk`: https://github.com/django/django/blob/main/django/db/models/base.py#L593-L601

File: api/api/models/media.py [hunk 3]
```diff
@@ -361,6 +363,22 @@ class AbstractMediaDecisionThrough(models.Model):
...
+        """Perform the action specified in the decision."""
+
+        if self.decision.action in {
+            DecisionAction.DEINDEXED_SENSITIVE,
+            DecisionAction.DEINDEXED_COPYRIGHT,
+        }:
+            self.deleted_media_class.objects.create(media_obj=self.media_obj)
+
+        if self.decision.action == DecisionAction.MARKED_SENSITIVE:
+            self.sensitive_media_class.objects.create(media_obj=self.media_obj)
```
Comment (sarayourfriend) [inline api/api/models/media.py:376]: Using `self.media_obj` will query the database, and isn't necessary here. (`_perform_es_action` queries it anyway to get the database identifier [truly an absurd and confusing distinction I will never understand why we don't do everything we can to get rid of], but we don't need to repeat bad patterns/habits and there it's unavoidable, unlike here where we can easily avoid it.)

```suggestion
        if self.decision.action in {
            DecisionAction.DEINDEXED_SENSITIVE,
            DecisionAction.DEINDEXED_COPYRIGHT,
        }:
            self.deleted_media_class.objects.create(media_obj_id=self.media_obj_id)

        if self.decision.action == DecisionAction.MARKED_SENSITIVE:
            self.sensitive_media_class.objects.create(media_obj_id=self.media_obj_id)
```

File: api/api/models/media.py [hunk 4]
```diff
@@ -361,6 +363,22 @@ class AbstractMediaDecisionThrough(models.Model):
     class Meta:
         abstract = True
 
+    def perform_action(self):
+        """Perform the action specified in the decision."""
+
+        if self.decision.action in {
```
Comment (sarayourfriend) [inline api/api/models/media.py:369]: A suggestion to avoid the implicit `self.decision` query when calling `perform_action` directly (if `save` doesn't run). This will be necessary when we implement the bulk decisions anyway.

```suggestion
    def perform_action(self, action=None):
        """Perform the action specified in the decision."""
        
        action = self.decision.action if action is None else action
        
        if action in {
```

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
The blocking change is to fix the issue with non-performant bulk decision creation.

I think we should exclude fixing the backfill from this PR and instead just fix the media admin so that it's working again. And then address the backfill (and lay groundwork for #3840, which needs this anyway) in a separate PR.

**Review by obulat — DISMISSED** [review]
After you delete a media object using a `deindexed_[...]` action, admin is trying to open the same `change_view` form with the same media object. However, because this media object has been deleted, you get an error on line `if tags := media_obj.tags:`. To prevent it, I added the redirect to the `change_view`:
```
        media_obj = self.get_object(request, object_id)
        if media_obj is None:
            return redirect(f"admin:api_{self.media_type}_changelist")
```

If you do not select any report, and submit a decision (e.g., `mark_sensitive`), you get no indication of the error except for the warning in the logs (which the moderator using the Admin UI will probably not see). This can be a follow-up issue since this PR is critical, but we should add the error display ("report_id" is required) to the form.

File: api/api/admin/media_report.py [hunk 5]
```diff
@@ -440,6 +440,10 @@ def decision_create_view(self, request, object_id):
         if request.method == "POST":
             media_obj = self.get_object(request, object_id)
 
+            through_model = {
+                "image": ImageDecisionThrough,
+                "audio": AudioDecisionThrough,
+            }[self.media_type]
```
[the remark this answers is not in the document]
Comment (dhruvkb) [reply api/api/admin/media_report.py:446]: The reason I write use the `= {}[]` pattern is because it is the most compact among all the alternatives and also raises an exception when none of the keys match the input.

I've done this a few times in this file itself. Would you prefer I change this pattern across the entire file, or keep this in in the interest of consistency, or just change it here?

File: api/api/models/media.py (hunk 4 above)
Comment (dhruvkb) [reply api/api/models/media.py]: Updated in fb8495f.

File: api/api/models/media.py (hunk 3 above)
Comment (dhruvkb) [reply api/api/models/media.py]: Updated in fb8495f.

File: api/api/management/commands/backfillmoderationdecision.py (hunk 1 above)
Comment (dhruvkb) [reply api/api/management/commands/backfillmoderationdecision.py]: I've reverted the changes to the backfill command to implement this separately.

File: api/api/admin/media_report.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply api/api/admin/media_report.py:450]: If the dict-key-access approach is preferred for any reason at all (it is nice it implicitly throws if the key doesn't exist), I'd at the very least think defining the dictionary outside the runtime scope of the function is a reasonable requirement, if only so that it's in a shared location. It's far-fetched to me to establish a pattern of defining otherwise static dictionaries, especially one encoding relationships between static objects, entirely dynamically in the runtime of a function. From a performance perspective it's fine here, but in a tight loop it's just silly, right? From a shared data/relationship encoding perspective (and discoverability, clarity, etc) it's definitely the worst option I can think of :confused:. Just seems like an antipattern to me :shrug: I also don't think compactness is necessarily a virtue, and certainly not in Python, which actively resists compactness in my experience.

**Comment by sarayourfriend** [comment]
@stacimc pointed out that the backfill doesn't need to "perform" the action at all, because it's just creating the decisions for actions that have already been performed. Glad we removed it already! I'll re-review this today.

**Review by sarayourfriend — APPROVED** [review]
LGTM, tests fine locally and I'm unblocking on my requested changes to get the fix into the admin.
